experimental/torch_xla2/torch_xla2/ops/jaten.py

```diff
@@ -64,6 +65,49 @@ def _handle_int64_trig(self, func):
   return res
 
 
+@op(torch.ops._c10d_functional.all_gather_into_tensor)
+def _c10d_all_gather(input, group_size: int, group_name: str):
+  return jax.lax.all_gather(input, "torch_dist")
+
+
+# TODO(wcromar): move c10d ops since they're not actually aten
+@op(torch.ops._c10d_functional.all_reduce)
+def _c10d_all_reduce(self, reduceOp: str, group_name: str):
+  match reduceOp:
+    case 'sum':
+      res = jax.lax.psum(self, axis_name="torch_dist")
+    case 'avg':
+      res = jax.lax.pmean(self, axis_name="torch_dist")
+    case 'min':
+      res = jax.lax.pmin(self, axis_name="torch_dist")
+    case 'max':
+      res = jax.lax.pmax(self, axis_name="torch_dist")
+    case _:
+      raise RuntimeError(f"Reduce op {reduceOp} not implemented")
+  return res
+
+
+@op(torch.ops._c10d_functional.broadcast)
+def _c10d_broadcast(self, src: int, group_name: str):
+  masked = jnp.where(
+      jax.lax.axis_index("torch_dist") == src,
+      self,
+      jnp.zeros_like(self),
+  )
+  return jax.lax.psum(masked, "torch_dist")
+
+
+@op(torch.ops._c10d_functional.wait_tensor)
+def _wait_tensor(tensor):
+  return tensor
```

Comment:
qq why not .wait here?

Reply:
To be honest, I don't actually understand why this works either.

There's no wait here because this is a JAX implementation, so some part of unwrapping the `Tensor` into an `Array` must get rid of the async wrapper. That's something I want to revisit. The dispatcher should probably be aware of `AsyncCollectiveTensor`s and `wait` them so our op implementation don't have to care (see how `copy_` has to have special handling right now).

Reply:
The dispatcher waits during `t2j` before calling a JAX implementation.